include/tvm/relax/expr.h

```diff
@@ -985,6 +985,11 @@ constexpr const char* kComposite = "Composite";
 constexpr const char* kPartitionedFromPattern = "PartitionedFromPattern";
 /*! \brief The required workspace for an external function. */
 constexpr const char* kWorkspaceSize = "WorkspaceSize";
+/*! \brief Indicate whether the function is pure (has no visible side effects for any input). */
+constexpr const char* kIsPure = "IsPure";
```

Comment:
my understanding is that is_pure is more like an annotation of the function, like ret_struct_info, it might make sense to make it directly as an attribution of the function (`Optional<Bool> purity_annotation`). Can be done as a followup

Reply:
I agree, actually. I brought up the issue before and at that time we didn't want to change the constructor. Maybe it's a better idea to do it in this PR to avoid even more disruption later.

Reply:
The main issue is how parsing would work. I'll see if I'll be able to do something about that

Reply:
I've implemented this suggestion. Please have a look at how it's turned out.